**Ejercicio de Clase: Generación de datos y redimensionado**


In [1]:
# Se importa la librería NumPy para poder utilizar sus funciones.
import numpy as np

# Paso 1: Crea un arreglo x con np.arange(1, 17) y redimensionalo en una matriz de 4x4.
# La función np.arange() genera una secuencia de números.
# La función reshape() cambia la estructura del arreglo sin modificar sus datos.
arreglo_x = np.arange(1, 17)
matriz_x = arreglo_x.reshape(4, 4)
print("Paso 1: Matriz 4x4 creada con arange y reshape:")
print(matriz_x)
print("-" * 50)

# Paso 2: Genera un arreglo de 10 elementos equidistantes entre 0 y 100 usando np.linspace().
# np.linspace() genera un vector con valores equidistantes en un rango.
arreglo_linspace = np.linspace(0, 100, 10)
print("Paso 2: Arreglo de 10 elementos entre 0 y 100:")
print(arreglo_linspace)
print("-" * 50)

# Paso 3: Crea una matriz aleatoria de 3x3 con enteros entre 1 y 20 usando np.random.randint().
# np.random.randint() genera una matriz con números enteros aleatorios.
matriz_aleatoria = np.random.randint(1, 21, size=(3, 3))
print("Paso 3: Matriz aleatoria 3x3:")
print(matriz_aleatoria)
print("-" * 50)

# Paso 4: Aplica np.sqrt() y np.log() sobre las matrices creadas.
# NumPy proporciona funciones matemáticas para aplicar a cada elemento de un arreglo.
print("Paso 4: Aplicación de funciones matemáticas:")

# Sobre la matriz de enteros
raiz_matriz_x = np.sqrt(matriz_x)
log_matriz_x = np.log(matriz_x)
print("\nRaíz cuadrada de la matriz del paso 1 (devuelve floats):")
print(raiz_matriz_x)
print("\nLogaritmo de la matriz del paso 1 (devuelve floats):")
print(log_matriz_x)

# Sobre el arreglo de floats (linspace)
# Nota: np.log(0) resultará en -inf y generará una advertencia (lo cual es esperado).
raiz_linspace = np.sqrt(arreglo_linspace)
log_linspace = np.log(arreglo_linspace)
print("\nRaíz cuadrada del arreglo del paso 2:")
print(raiz_linspace)
print("\nLogaritmo del arreglo del paso 2 (con -inf para log(0)):")
print(log_linspace)
print("-" * 50)

Paso 1: Matriz 4x4 creada con arange y reshape:
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]
--------------------------------------------------
Paso 2: Arreglo de 10 elementos entre 0 y 100:
[  0.          11.11111111  22.22222222  33.33333333  44.44444444
  55.55555556  66.66666667  77.77777778  88.88888889 100.        ]
--------------------------------------------------
Paso 3: Matriz aleatoria 3x3:
[[10  2  5]
 [ 6  8  9]
 [ 4 16 19]]
--------------------------------------------------
Paso 4: Aplicación de funciones matemáticas:

Raíz cuadrada de la matriz del paso 1 (devuelve floats):
[[1.         1.41421356 1.73205081 2.        ]
 [2.23606798 2.44948974 2.64575131 2.82842712]
 [3.         3.16227766 3.31662479 3.46410162]
 [3.60555128 3.74165739 3.87298335 4.        ]]

Logaritmo de la matriz del paso 1 (devuelve floats):
[[0.         0.69314718 1.09861229 1.38629436]
 [1.60943791 1.79175947 1.94591015 2.07944154]
 [2.19722458 2.30258509 2.39789527 2.48490665]
 [2.

C:\Users\hcr88\AppData\Local\Temp\ipykernel_7840\3810592683.py:42: RuntimeWarning: divide by zero encountered in log
  log_linspace = np.log(arreglo_linspace)


**<p style="font-size:18px; color:auto;">Análisis: Aplicación de Funciones (Paso 5)</p>**

¿Qué diferencias notas en la forma de aplicar funciones sobre arreglos enteros vs flotantes?

La diferencia fundamental no está en cómo se aplica la función (la sintaxis es idéntica), sino en el tipo de dato del resultado.

* **Arreglos de Enteros:** Cuando aplicas una función matemática como np.sqrt() o np.log() a un arreglo de enteros, NumPy realiza la operación y devuelve un nuevo arreglo de tipo flotante (float). Esto es necesario porque el resultado de estas operaciones raramente es un número entero (por ejemplo, sqrt(2) es 1.414...). NumPy maneja esta conversión de tipo de dato automáticamente.

* **Arreglos de Flotantes:** Cuando la misma función se aplica a un arreglo que ya contiene flotantes (como el generado por np.linspace), el resultado es, como es de esperar, otro arreglo de flotantes.

En resumen, la principal diferencia es la conversión de tipo implícita que NumPy realiza al operar sobre enteros para no perder la precisión decimal del resultado.